### ISSUE TITLE
Potential bugged MKL lib included in conda 5.0

### ISSUE THREAD

**Issue by fatrabber** [ContinuumIO/anaconda-issues#7480]
This issue is redirected from https://github.com/numpy/numpy/issues/10071

Here is what I think may help to reproduce the bug:
1. Install anaconda 5.0, which has newest MKL. (My OS is Centos 7)
2. Install tensorflow, which creates an environment that uses newest MKL. 

The way I installed was: 
	pip install tensorflow
	pip install tensorflow-gpu
	
	conda list | grep tensor 
	tensorflow                1.3.0                         0  
	tensorflow-base           1.3.0            py27h0dbb4d0_1  
	tensorflow-gpu            1.3.0                         0  
	tensorflow-gpu-base       1.3.0           py27cuda8.0cudnn6.0_1  
	tensorflow-tensorboard    0.1.5                    py27_0  

3. Run this program:

```
import numpy as np
import tensorflow as tf
row, col = 76870, 3073
#row, col = 10, 3073
a = np.random.normal(size=(row, col))
b = np.random.normal(size=(col, 1))
np_outer = np.dot(a,b)
print(np_outer[0,0])
old_way = 0.0
for i in range(col):
    old_way = old_way + a[0,i] * b[i,0]

print(old_way)
```

Two outputs will be different.

There are few tricks here:
1. If "import tensorflow as tf" is commented out, results are fine. I think this is because tf uses the news MKL.
2. If the data is not enough, say row is 10 instead of 76870, the results are fine. I think this is because MKL will only be used when the data is large.

If more people can reproduce this, I think we can be sure it's a problem of the newest MKL lib.

Thanks,

**Comment by mingwandroid** [comment]
> Install anaconda 5.0, which has newest MKL

This is not correct. `Anaconda 5.0.1` (which I assume you meant) comes with `mkl 2018.0` whereas the latest released version is `mkl 2018.1`. You should run `conda update mkl` to get this.

> If more people can reproduce this, I think we can be sure it's a problem of the newest MKL lib.

No, that does not follow at all. Using `pip` and `conda` in the same environment leads to conflicts, particularly for things involving C or C++ libraries such as `tensorflow`. Is there any reason not to use `conda install tensorflow-gpu` here?

**Comment by fatrabber** [comment]
Install anaconda 5.0, which has newest MKL
	This is not correct. Anaconda 5.0.1 (which I assume you meant) comes with mkl 2018.0 whereas the latest released version is mkl 2018.1. You should run conda update mkl to get this.

My bad. This is the newest MKL I could used so I assume it's newest. You are right, the version I am using is 2018.0.1 as showed below:

	conda list | grep -i mkl
	mkl                       2018.0.1             h19d6760_4  
	mkl-service               1.1.2            py27hb2d42c5_4  

For some reason, I cannot update my mkl to 2018.1 either:

	conda update mkl
	Fetching package metadata ...........
	Solving package specifications: .

	# All requested packages already installed.
	# packages in environment at /opt/anaconda2:
	#
	mkl                       2018.0.1             h19d6760_4  
--	

	If more people can reproduce this, I think we can be sure it's a problem of the newest MKL lib.
	No, that does not follow at all. Using pip and conda in the same environment leads to conflicts, particularly for things involving C or C++ libraries such as tensorflow. Is there any reason not to use conda install tensorflow-gpu here?
	
I was certainly confused with pip install and conda install, which I thought were same. I just double-checked my .zsh_history and what I used were conda install

	cat .zsh_history | grep tensorflow
	: 1509504981:0;conda install tensorflow
	: 1509517145:0;conda install --upgrade tensorflow-gpu
	: 1509517152:0;conda install tensorflow-gpu

I am not familiar with conda's virtualenv, but in my opinion we can construct a test like this:
	1. Set up two env: env_a has MKL 2018.0,1 and env_b has no MKL.
	2. Run my program in two env and compare the results. If in env_b two outputs are the same and in env_a two outputs are different, I think we can say it's likely MKL 2018.0.1's problem. As the number of machines that can reproduce increases, our confidence can be incresed too.
	
In my program, I just imported tensorflow or not to try to construct env_a and env_b. But certainly I cannot make sure all other libs are the same.

Best,

**Comment by fatrabber** [comment]
To pinpoint the bug, I created a new environment and downgraded MKL from 2018.0.1 to 2017.0.4. Several other libs were downgraded too:

    mkl:          2018.0.1-h19d6760_4   --> 2017.0.4-h4c4d0af_0
    mkl-service:  1.1.2-py27hb2d42c5_4  --> 1.1.2-py27_3
    numexpr:      2.6.2-py27h9521900_2  --> 2.6.2-np113py27_0
    numpy:        1.13.3-py27hbcc08e0_0 --> 1.13.1-py27_0
    scikit-learn: 0.19.1-py27h445a80a_0 --> 0.19.0-np113py27_0
    scipy:        1.0.0-py27hf5f0f52_0  --> 0.19.1-np113py27_0
    
The results look good in this environment. If I switch back to the env with MKL 2018.0.1 and numpy 1.13.3, the program will have bug (results will be different).

Unfortunately mkl 2018.0.1 and nupy 1.13.3 are bound together, so I cannot figure out if the bug is from mkl 2018.0.1 or numpy 1.13.3. But I think from here I can be sure one of them is bugged.

For now I will downgrade my mkl from 2018.0.1 to 2017.0.4. Hopefully newer version of MKL will be released and the bug will be fixed.




P.S. Previously I thought tensorflow or large data triggers a different environment that includes MKL 2018.0.1 that has bug, but it turns out that the bug was not environment related, as I had only one environment. I guess it's optimization option related, which I have no idea how it works.

**Comment by msarahan** [comment]
Thanks for helping to isolate the issue further.  I have asked our Intel contacts, @oleksandr-pavlyk and @tomashek to take a look.

**Comment by oleksandr-pavlyk** [comment]
I tried to reproduce the problem using the following environment:

```
# packages in environment at /tmp/miniconda3/envs/anac5_py3:
#
backports                 1.0              py36hfa02d7e_1  
backports.weakref         1.0rc1                   py36_0  
bleach                    1.5.0                    py36_0  
ca-certificates           2017.08.26           h1d4fec5_0  
certifi                   2017.11.5        py36hf29ccca_0  
html5lib                  0.9999999                py36_0  
intel-openmp              2018.0.0             hc7b2577_8  
libedit                   3.1                  heed3624_0  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 7.2.0                h7cc24e2_2  
libprotobuf               3.4.1                h5b8497f_0  
libstdcxx-ng              7.2.0                h7a57d05_2  
markdown                  2.6.9                    py36_0  
mkl                       2018.0.1             h19d6760_4  
ncurses                   6.0                  h9df7e31_2  
numpy                     1.13.3           py36ha12f23b_0  
openssl                   1.0.2m               h26d622b_1  
pip                       9.0.1            py36h6c6f9ce_4  
protobuf                  3.4.1            py36h306e679_0  
python                    3.6.3                h6c0c0dc_5  
readline                  7.0                  ha6073c6_4  
setuptools                36.5.0           py36he42e2e1_0  
six                       1.11.0           py36h372c433_1  
sqlite                    3.20.1               hb898158_2  
tensorflow                1.3.0                         0  
tensorflow-base           1.3.0            py36h5293eaa_1  
tensorflow-tensorboard    0.1.5                    py36_0  
tk                        8.6.7                hc745277_3  
werkzeug                  0.12.2           py36hc703753_0  
wheel                     0.30.0           py36hfd4bba0_1  
xz                        5.2.3                h55aa19d_2  
zlib                      1.2.11               ha838bed_2  
```

and could not reproduce the discrepancy between `old_way` and `np_outer[0,0]`, using the provided reproducing script, and running on ``CentOS 7.3.1611``.

After executing `conda install tensorflow-gpu`, the script stopped running with the error:

```
ImportError: libcuda.so.1: cannot open shared object file: No such file or directory

Failed to load the native TensorFlow runtime.
```

**Comment by oleksandr-pavlyk** [comment]
@fatrabber Could you please rerun the script in your environment like so `MKL_THREADING_LAYER=GNU python script.py`, and like `MKL_THREADING_LAYER=sequential python script.py` to check if any of these runs result in identical results of `old_way` and `np_outer[0,0]`. Thanks.

**Comment by fatrabber** [comment]
I run in both your ways and the results are good:

(newest_mkl_2018) ➜  tst python bad_mkl.py
-524.234161009
-87.3723601682
(newest_mkl_2018) ➜  tst MKL_THREADING_LAYER=GNU python bad_mkl.py
-20.5544152819
-20.5544152819
(newest_mkl_2018) ➜  tst MKL_THREADING_LAYER=sequential python bad_mkl.py
25.9515522148
25.9515522148

**Comment by oleksandr-pavlyk** [comment]
@fatrabber Thank you for checking. Essentially, the issue is a component in your application, likely `tensorflow` is linked against Gnu OpenMP. Intel(R) MKL by default uses Intel(R) Open MP as a threading layer. Both libraries implement common symbols, so mixing them is not a good idea, as the loading order and OS-specific peculiarities of shared library loader may cause a symbol intended to be resolved to Intel OMP be actually loading Gnu OMP.

In your particular application Gnu OMP is loaded first, and it is causing the problem in MKL with default setting. Should Intel OMP be loaded first, this may cause yet to be identified issues with the application intended to work with Gnu OMP.

Using `MKL_THREADING_LAYER=GNU` ensures consistency of OMP implementation across your entire Python application, while using `MKL_THREADING_LAYER=sequential` inhibits threading in MKL, and therefore the problem does not arise.

**Comment by msarahan** [comment]
@jjhelmus - let's talk about how to address this with tensorflow.  Setting MKL_THREADING_LAYER is  quite a headache, since it can really only be set once, and it's not clear to me who should set it, and how.  The Theano team ran into this same issue, but I'm not sure if they ever came up with a solution they liked.

**Comment by jjhelmus** [comment]
Our `tensorflow-gpu-base` package does link against the GNU OMP library which seems to be consist with the results seen.  I do not believe it is possible avoid this linkage although I'm happy to investigate.

This looks to effect only the `tensorflow-gpu-base` version of tensorflow, the non-gpu version, `tensorflow-base`, does not link against the GNU OMP library and therefore should not be effected.  

Additionally, if MKL is not installed in the environment the results seem to be consistent.  After creating an environment using `conda create -n tf_gpu_nomkl python=2.7 tensorflow-gpu nomkl` to select the NumPy package linked against OpenBLAS, the script provided by @fatrabber produced good results.

**Comment by fatrabber** [comment]
Thanks for pointing out the problem. 

My another question is: how do MKL 2017.0.4 and previous versions work and what is different between MKL 2018.0.1 and the versions before? 

As I tried, after I downgraded MKL from 2018.0.1 to 2017.0.4, the problem went away. If it's something changed that triggers the linking problem, I wonder if we can use the previous way in MKL 2018.0.1.

**Comment by oleksandr-pavlyk** [comment]
@fatrabber Note that downgrading MKL to 2017.0.4 also caused NumPy to downgrade to 1.13.1. The difference between 1.13.3 and 1.13.1, in what concerns this behavior, can be seen in the output of `np.__config__.show()`. In 1.13.1 it would show multiple libraries linked, such as `mkl_intel_lp64`, `iomp5`, etc, while in 1.13.3 it will show only `mkl_rt`. 

The `mkl_rt` library is a dynamic loader, which features ability to [dynamically select](https://software.intel.com/en-us/mkl-macos-developer-guide-dynamically-selecting-the-interface-and-threading-layer) threading layer, among other things. Depending the dynamically selected choices this library loads respective implementing shared libraries. 

By default MKL is  configured to use Intel Open(R) MP threading layer, and so MKL would load `iomp5` library to use its functions to implement threading.

What transpires next has to do with how OS resolves symbols to these libraries and the fact that both Gnu OMP and Intel OMP implement the same standard, and hence their functions have the same name.
Shared library loader may decide not to load the library MKL expects (iomp5) because the symbol it needs from it has already been loaded from Gnu OMP (such a symbol may be `set_omp_num_threads`, for instance). Other symbols may be loaded from iomp5, as expected by MKL. 

The solution may be as simple as to set `MKL_THREADING_LAYER` in you script prior to loading of any libraries:

```python
import os
os.environ['MKL_THREADING_LAYER'] = 'GNU'
import numpy as np
import tensorflow as tf
row, col = 76870, 3073
#row, col = 10, 3073
a = np.random.normal(size=(row, col))
b = np.random.normal(size=(col, 1))
np_outer = np.dot(a,b)
print(np_outer[0,0])
old_way = 0.0
for i in range(col):
    old_way = old_way + a[0,i] * b[i,0]

print(old_way)
```

Could you please confirm whether it works?

Thanks

**Comment by fatrabber** [comment]
The new program works well. Thanks for explaining. I always thought the problem is introduced by MKL's change, but it's actually from numpy.

I agree with @msarahan that understanding such linking and setting threading layer are probably too much for dev and researchers like us. It would be great if the linking is done at some lower level and programmers do not need to worry about it.

For now, my solution will be staying with numpy 1.13.1. If someone is using 1.13.3, I guess it is safer to always set ['MKL_THREADING_LAYER'] = 'GNU' when numpy is imported. Right?

**Comment by oleksandr-pavlyk** [comment]
@fatrabber No, not always. The choice does not matter is none of the native extensions in your python stack explicitly link against any OpenMP threading libraries.

Like you have a native extensions that is explicitly linked against Gnu OMP in your stack, which necessitates use of GNU threading layer for consistency, you may similarly have a native extensions explicitly linking against Intel OMP, in which case setting `MKL_THREADING_LAYER` to `GNU` will cause problems.

I guess the lesson is the choice of threading layers in your stack best be consistent, but at least should not employ different realizations of the same standard.

**Comment by fatrabber** [comment]
Thanks for the explanation.

I guess my next question would be how to find out which threading lib a python lib is linking to, in case we have to set the threading layer by ourselves.

**Comment by oleksandr-pavlyk** [comment]
On Linux, this would be by using `ldd`, for example:

```
ldd $CONDA_PREFIX/lib/python*/site-packages/scipy/optimize/_cobyla*.so 
```

with the output being:

```
        linux-vdso.so.1 =>  (0x00007fff000df000)
        libgfortran.so.4 => <elided>/lib/python3.5/site-packages/scipy/optimize/../../../../libgfortran.so.4 (0x00002b0421135000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00002b042145f000)
        libquadmath.so.0 => <elided>/lib/python3.5/site-packages/scipy/optimize/../../../../libquadmath.so.0 (0x00002b0421824000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00002b0421a55000)
        libgcc_s.so.1 => <elided>/lib/python3.5/site-packages/scipy/optimize/../../../../libgcc_s.so.1 (0x00002b0421d5b000)
        /lib64/ld-linux-x86-64.so.2 (0x00002b0420cf1000)
```

**Comment by fatrabber** [comment]
Not sure if I understand correctly. Let me try to put the workflow together:
1. ldd a random .so file under the folder of python lib that will be used
2. check keywords like "GNU" or "INTEL" to find out what threading layer the python lib is using
3. set the threading layer to that in the python program

Here's what I have tried to practice:

You mentioned "tensorflow is linked against Gnu OpenMP. Intel(R) MKL by default uses Intel(R) Open MP as a threading layer." So I tried "ldd"ing tensorflow and mkl which used by numpy, while I found no keywords like "GNU", "INTEL".

(Both numpy and tensorflow links to /lib64/libpthread.so.0. Not sure if this is related tho.)

Here are my confusions: 
1. If we do see "GNU" or "INTEL", we have to set threading layer to that.
2. I suppose we don't see any "GNU" or "INTEL" keywords because the linking is default, so setting the layer to either GNU or Intel would work. 

Are they correct?


    ➜  numpy pwd
    /opt/anaconda2/envs/newest_mkl_2018/lib/python2.7/site-packages/numpy
    ➜  numpy ldd ./core/multiarray.so
        linux-vdso.so.1 =>  (0x00007ffc6cd70000)
        libmkl_rt.so => /opt/anaconda2/envs/newest_mkl_2018/lib/python2.7/site-packages/numpy/./core/../../../../libmkl_rt.so (0x00007fbb05dda000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fbb05ba3000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fbb058a1000)
        libpython2.7.so.1.0 => /opt/anaconda2/envs/newest_mkl_2018/lib/python2.7/site-packages/numpy/./core/../../../../libpython2.7.so.1.0 (0x00007fbb054c4000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fbb05100000)
        /lib64/ld-linux-x86-64.so.2 (0x000055e876c46000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fbb04efc000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00007fbb04cf8000)

    ➜  lib pwd
    /opt/anaconda2/envs/newest_mkl_2018/lib
    ➜  lib ldd libmkl_rt.so
        linux-vdso.so.1 =>  (0x00007fffdd743000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fefbad06000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fefba942000)
        /lib64/ld-linux-x86-64.so.2 (0x00005611e2277000)
        
    ➜  tensorflow pwd
    /opt/anaconda2/envs/newest_mkl_2018/lib/python2.7/site-packages/tensorflow
    ➜  tensorflow ldd ./contrib/batching/python/ops/_batch_ops.so
        linux-vdso.so.1 =>  (0x00007fffc6b7b000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f01267d2000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f01265b5000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f01263b1000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f01261a9000)
        libstdc++.so.6 => /opt/anaconda2/envs/newest_mkl_2018/lib/python2.7/site-packages/tensorflow/./contrib/batching/python/ops/../../../../../../../libstdc++.so.6 (0x00007f0125e6e000)
        libgcc_s.so.1 => /opt/anaconda2/envs/newest_mkl_2018/lib/python2.7/site-packages/tensorflow/./contrib/batching/python/ops/../../../../../../../libgcc_s.so.1 (0x00007f0125c5c000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f0125899000)
        /lib64/ld-linux-x86-64.so.2 (0x00005591788ca000)

**Comment by lorenzoridolfi** [comment]
I'm also with an issue with np.dot

I made this script:

import numpy as np
from numba import jit
row, col = 76870, 10

@jit
def dot(a,b):
	n = a.shape[0]
	m = a.shape[1]
	out = np.zeros(n)
	for i in range(n):
		sum = 0
		for j in range(m):
			sum += a[i,j] * b[j]
		out[i] = sum
	return out

a = np.random.normal(size=(row, col))
b = np.random.normal(size=(col, 1))

np_dot = np.dot(a,b)
dot    = dot(a,b)
print(np.linalg.norm(np_dot - dot))

and the result should be zero but is 22913.2542827

**Comment by msarahan** [comment]
This should now be fixed in numpy 1.13.3, build 2, or numpy 1.14.0, build 1.

The real fix will be in later MKL versions, but these numpy builds include this fix to detect which openmp implementation should be used: https://github.com/AnacondaRecipes/numpy-feedstock/pull/5

Thanks @oleksandr-pavlyk for helping to diagnose and fix this.
